smith.sh

```diff
@@ -299,14 +306,73 @@ function rollout_status()
 {
   if [ -n "${PROJECT_NAME_VAR}" ];
   then
-    kubectl rollout status deployment ${PROJECT_NAME_VAR}
+
+    interval=1 #1 second
+    rollout_status_raw & CPID=$!
+
+    ((t = $DEPLOY_TIMEOUT_VAR ))
+
+      while ((t > 0));
+        do
+          if [ -n "$(ps -p ${CPID} -o pid=)" ]
+            then
+              sleep $interval;
+              ((t -= $interval))
+            else
+              exit 0;
+          fi
+        done
+
+    rollout_status_error $CPID;
+
   else
     echo "ERROR: needs project_name - $0 -h for usage";
   fi;
   exit;
 }
 
 
+function rollout_status_raw()
+{
+  kubectl rollout status deployment ${PROJECT_NAME_VAR}
+}
+
+
+function rollout_status_error()
+{
+  if [ -n "$(ps -p $1 -o pid=)" ]
+    then
+      rollout_status_error_info;
+      echo "INFO: Killing deployment process as is timed out";
+      kill -s SIGTERM $1 && kill -0 $1 || exit 0
+      sleep 5;
+      kill -s SIGKILL $1
+    else
+     exit;
+fi;
+}
+
+
+function rollout_status_error_info(){
+
+  kubectl get pods --selector=app=${PROJECT_NAME_VAR} --no-headers=true |  \
```

Comment:
consider using `go-template` output instead of all this logic

`kubectl get pods --selector=app=camversity-session -o go-template='{{range .items}}{{if ne .status.phase "Running"}} {{.metadata.name}} {{.status.phase}}{{"\n"}}{{end}}{{end}}'`

Reply:
You only get rid of if condition but that can be done with grep -v .

Reply:
You get rid of that horrible while loop and inline exec & awk

Reply:
go template on status.phase returns false status:

[root@aneamtu-nix cli]# kubectl get pods --selector=app=camversity-crashtesting -o go-template='{{range .items}}{{.metadata.name}} {{.status.phase}}{{"\n"}}{{end}}'
camversity-crashtesting-6987c7d9f5-6zlln Running
camversity-crashtesting-6cbdbcb6d7-thl75 Running
[root@aneamtu-nix cli]# kubectl get pods --selector=app=camversity-crashtesting
NAME                                       READY     STATUS             RESTARTS   AGE
camversity-crashtesting-6987c7d9f5-6zlln   0/1       CrashLoopBackOff   11         31m
camversity-crashtesting-6cbdbcb6d7-thl75   1/1       Running            0          1d

It needs docker container status reason:
kubectl get pods --selector=app=camversity-crashtesting -o go-template='{{range $index, $element:= .items}}{{range .status.containerStatuses}}{{if .state.waiting.reason }}{{$element.metadata.name}}:{{ .name }}:{{.state.waiting.reason }}{{"\n"}}{{end}}{{end}}{{end}}'

Since we could have more than one pod and also containers are lists, we still need some loop.